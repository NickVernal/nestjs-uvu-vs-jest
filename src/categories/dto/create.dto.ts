import { ApiProperty } from '@nestjs/swagger';
import { IsOptional, IsString, MinLength } from 'class-validator';
import { Category } from '../categories.entity';

export class CreateCategoryRequestData {
  @ApiProperty()
  @IsString()
  @MinLength(8)
  slug: string;

  @ApiProperty()
  @IsString()
  @MinLength(3)
  name: string;

  @ApiProperty({ required: false })
  @IsOptional()
  @IsString()
  description?: string;
}

export class CreateCategoryResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';

  @ApiProperty({ type: Category })
  payload: Category;
}
