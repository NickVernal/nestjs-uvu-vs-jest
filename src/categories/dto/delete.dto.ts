import { ApiProperty } from '@nestjs/swagger';
import { IsUUID } from 'class-validator';

export class DeleteCategoryRequestParams {
  @ApiProperty()
  @IsUUID(4)
  id: string;
}

export class DeleteCategoryResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';
}
