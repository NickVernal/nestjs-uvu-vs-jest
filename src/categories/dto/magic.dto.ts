import { ApiProperty } from '@nestjs/swagger';
import { IsUUID } from 'class-validator';
import { Category } from '../categories.entity';

export class MagicRequestParams {
  @ApiProperty()
  @IsUUID(4)
  id: string;
}

export class MagicResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';

  @ApiProperty({ type: Category })
  payload: Category;
}
