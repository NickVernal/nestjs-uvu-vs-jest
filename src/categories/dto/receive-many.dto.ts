import { ApiProperty } from '@nestjs/swagger';
import { Category } from '../categories.entity';

export class ReceiveManyResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';

  @ApiProperty({ type: Category, isArray: true })
  payload: Category[];
}
