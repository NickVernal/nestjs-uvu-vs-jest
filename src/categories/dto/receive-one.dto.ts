import { ApiProperty } from '@nestjs/swagger';
import { IsUUID } from 'class-validator';
import { Category } from '../categories.entity';

export class ReceiveOneRequestParams {
  @ApiProperty()
  @IsUUID(4)
  id: string;
}

export class ReceiveOneResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';

  @ApiProperty({ type: Category })
  payload: Category;
}
