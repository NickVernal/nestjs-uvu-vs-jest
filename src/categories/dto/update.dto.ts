import { ApiProperty } from '@nestjs/swagger';
import { IsOptional, IsString, IsUUID, MinLength } from 'class-validator';
import { Category } from '../categories.entity';

export class UpdateCategoryRequestParams {
  @ApiProperty()
  @IsUUID(4)
  id: string;
}

export class UpdateCategoryRequestData {
  @ApiProperty()
  @IsString()
  @MinLength(8)
  slug: string;

  @ApiProperty()
  @IsString()
  @MinLength(3)
  name: string;

  @ApiProperty({ required: false })
  @IsOptional()
  @IsString()
  description: string;
}

export class UpdateCategoryResponse {
  @ApiProperty({ example: 'ok' })
  status: 'ok';

  @ApiProperty({ type: Category })
  payload: Category;
}
