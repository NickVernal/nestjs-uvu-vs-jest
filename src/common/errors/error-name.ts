export enum ErrorName {
  CategoryDoesNotExistError = 'CATEGORY_DOES_NOT_EXIST_ERROR',
}
