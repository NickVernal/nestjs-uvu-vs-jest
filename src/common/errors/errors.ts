import { CustomError } from './custom-error';
import { ErrorName } from './error-name';

export class CategoryDoesNotExistError extends CustomError {
  constructor() {
    super(ErrorName.CategoryDoesNotExistError);
  }
}
