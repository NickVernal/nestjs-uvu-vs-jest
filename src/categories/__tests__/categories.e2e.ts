import { INestApplication } from '@nestjs/common';
import request from 'supertest';
import { v4 as uuid } from 'uuid';

import { apiKeyHeader } from '../../constants';
import { env } from '../../config';
import { routes } from '../../routes';
import {
  clearTables,
  dropConnections,
  setup,
  getCategories,
  createCategory,
} from '../../utils/tests';
import { createMask } from '../../utils/mask-fields';
import { buildRoute } from '../../utils/build-route';
import { ErrorName } from '../../common/errors/error-name';

import { CreateCategoryRequestData } from '../dto/create.dto';

let app: INestApplication;

beforeAll(async () => {
  app = await setup();
});

afterEach(async () => {
  await clearTables();
});

afterAll(async () => {
  await dropConnections();
  await app.close();
});

const mask = createMask(['id']);

describe('Categories:', () => {
  describe('Create category:', () => {
    const route = routes.root + routes.categories.create;
    it('No auth header -> 401', async () => {
      const response = await request(app.getHttpServer())
        .post(route)
        .send({
          description: '1234',
          name: '1234',
          slug: '1234',
        } as CreateCategoryRequestData);
      expect(response.statusCode).toEqual(401);
    });

    it('Wrong auth header -> 401', async () => {
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY + 'x')
        .send({
          description: '1234',
          name: '1234',
          slug: '1234',
        } as CreateCategoryRequestData);
      expect(response.statusCode).toEqual(401);
    });

    it('Correct auth header, no required data -> 400', async () => {
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(400);
      expect(response.body).toMatchSnapshot();
    });

    it('Correct auth header, only required fields -> 201', async () => {
      const dataToCreate: CreateCategoryRequestData = {
        name: '1234',
        slug: '12345678',
      };

      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY)
        .send(dataToCreate);

      expect(response.statusCode).toEqual(201);
      expect(mask(response.body)).toMatchSnapshot();
      expect(response.body?.payload).toMatchObject(dataToCreate);

      const [category] = await getCategories();
      expect(category).toMatchObject(dataToCreate);
    });

    it('Correct auth header, all fields -> 201', async () => {
      const dataToCreate: CreateCategoryRequestData = {
        name: '1234',
        slug: '12345678',
        description: '1234',
      };

      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY)
        .send(dataToCreate);

      expect(response.statusCode).toEqual(201);
      expect(mask(response.body)).toMatchSnapshot();
      expect(response.body?.payload).toMatchObject(dataToCreate);

      const [category] = await getCategories();
      expect(category).toMatchObject(dataToCreate);
    });
  });

  describe('Receive one category:', () => {
    const baseRoute = routes.root + routes.categories.receiveOne;

    it('No auth header -> 401', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer()).get(route);
      expect(response.statusCode).toEqual(401);
    });

    it('Wrong auth header -> 401', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .get(route)
        .set(apiKeyHeader, env.API_KEY + 'x');
      expect(response.statusCode).toEqual(401);
    });

    it('Correct auth header, no such category -> 400, CATEGORY_DOES_NOT_EXIST_ERROR', async () => {
      const id = uuid();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .get(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(400);
      expect(response.body.message).toEqual(
        ErrorName.CategoryDoesNotExistError,
      );
    });

    it('Correct auth header, category exists -> 200', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .get(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(200);
      expect(mask(response.body)).toMatchSnapshot();
    });
  });

  describe('Trigger magic:', () => {
    const baseRoute = routes.root + routes.categories.magic;

    it('No auth header -> 401', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer()).post(route);
      expect(response.statusCode).toEqual(401);
    });

    it('Wrong auth header -> 401', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY + 'x');
      expect(response.statusCode).toEqual(401);
    });

    it('Correct auth header, no such category -> 400, CATEGORY_DOES_NOT_EXIST_ERROR', async () => {
      const id = uuid();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(400);
      expect(response.body.message).toEqual(
        ErrorName.CategoryDoesNotExistError,
      );
    });

    it('Correct auth header, category exists, cant perform magic -> 400, CATEGORY_CAN_NOT_DO_MAGIC_ERROR', async () => {
      const { id } = await createCategory({ canDoMagic: false });
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(400);
      expect(response.body.message).toEqual(
        ErrorName.CategoryCanNotDoMagicError,
      );
    });

    it('Correct auth header, category exists -> 201', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });
      const response = await request(app.getHttpServer())
        .post(route)
        .set(apiKeyHeader, env.API_KEY);
      expect(response.statusCode).toEqual(201);
      expect(mask(response.body)).toMatchSnapshot();
      expect(response.body?.payload?.canDoMagic).toStrictEqual(false);

      const [category] = await getCategories();
      expect(mask(category as any)).toMatchSnapshot();
    });

    it('Race condition', async () => {
      const { id } = await createCategory();
      const route = buildRoute(baseRoute, { id });

      const responses = await Promise.all(
        new Array(10)
          .fill(0)
          .map(() =>
            request(app.getHttpServer())
              .post(route)
              .set(apiKeyHeader, env.API_KEY),
          ),
      );

      const failed = responses.filter(
        ({ statusCode, body }) =>
          statusCode === 400 &&
          body?.message === ErrorName.CategoryCanNotDoMagicError,
      );
      const [succeeded] = responses.filter(
        ({ statusCode }) => statusCode === 201,
      );
      expect(failed.length).toEqual(9);
      expect(mask(succeeded.body)).toMatchSnapshot();
    });
  });
});
