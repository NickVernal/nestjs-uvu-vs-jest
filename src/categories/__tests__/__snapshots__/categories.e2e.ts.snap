// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories: Create category: Correct auth header, all fields -> 201 1`] = `
Object {
  "payload": Object {
    "canDoMagic": true,
    "description": "1234",
    "id": "*",
    "name": "1234",
    "slug": "12345678",
  },
  "status": "ok",
}
`;

exports[`Categories: Create category: Correct auth header, no required data -> 400 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "slug must be longer than or equal to 8 characters",
    "slug must be a string",
    "name must be longer than or equal to 3 characters",
    "name must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`Categories: Create category: Correct auth header, only required fields -> 201 1`] = `
Object {
  "payload": Object {
    "canDoMagic": true,
    "description": "",
    "id": "*",
    "name": "1234",
    "slug": "12345678",
  },
  "status": "ok",
}
`;

exports[`Categories: Receive one category: Correct auth header, category exists -> 200 1`] = `
Object {
  "payload": Object {
    "canDoMagic": true,
    "description": "",
    "id": "*",
    "name": "Clothes",
    "slug": "clothes",
  },
  "status": "ok",
}
`;

exports[`Categories: Trigger magic: Correct auth header, category exists -> 201 1`] = `
Object {
  "payload": Object {
    "canDoMagic": false,
    "description": "",
    "id": "*",
    "name": "Clothes",
    "slug": "clothes",
  },
  "status": "ok",
}
`;

exports[`Categories: Trigger magic: Correct auth header, category exists -> 201 2`] = `
Category {
  "canDoMagic": false,
  "description": "",
  "id": "*",
  "name": "Clothes",
  "slug": "clothes",
}
`;

exports[`Categories: Trigger magic: Race condition 1`] = `
Object {
  "payload": Object {
    "canDoMagic": false,
    "description": "",
    "id": "*",
    "name": "Clothes",
    "slug": "clothes",
  },
  "status": "ok",
}
`;
